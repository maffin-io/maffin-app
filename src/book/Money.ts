import * as currencies from '@dinero.js/currencies';
import type { Currency as DineroCurrency } from 'dinero.js';
import * as djs from 'dinero.js';

import { toAmountWithScale, moneyToString } from '@/helpers/number';

export default class Money {
  private _raw: djs.Dinero<number>;

  /**
   * A wrapper class for dinero object to reduce verbosity and overhead. The constructor
   * checks for how many decimals the amount has and calculates the scale automatically.
   *
   * @param amount - The amount to be considered in the money object. Can be a float
   * @param currency - The currency
   * @param scale - Number representing custom scale. Default is 0, if passed differently, that one
   * is taken into account
   */
  constructor(amount: number, currency: string, scale?: number) {
    const dineroCurrency = toCurrency(currency);

    if (scale === undefined) {
      const { amount: newAmount, scale: newScale } = toAmountWithScale(amount);
      this._raw = djs.dinero({
        currency: dineroCurrency,
        amount: newAmount,
        scale: newScale,
      });
    } else {
      this._raw = djs.dinero({
        currency: dineroCurrency,
        amount,
        scale,
      });
    }
  }

  get currency(): string {
    return djs.toSnapshot(this._raw).currency.code;
  }

  get raw(): djs.Dinero<number> {
    return this._raw;
  }

  /**
   * Represents the money object as a string with the currency standardizing
   * the scale to 2 as a default. If the scale is passed then it uses that.
   */
  toString(scale = 2): string {
    return djs.toDecimal(
      djs.transformScale(this._raw, scale),
      ({ value, currency }) => `${value} ${currency.code}`,
    );
  }

  /**
   * Represents the money object as a string with the currency standardizing
   * the scale to 2 as a default. If the scale is passed then it uses that.
   *
   * The result is a localized string with the currency as a symbol
   */
  format(scale = 2): string {
    return djs.toDecimal(
      djs.transformScale(this._raw, scale),
      ({ value, currency }) => moneyToString(Number(value), currency.code),
    );
  }

  toNumber(): number {
    return parseFloat(djs.toDecimal(this._raw));
  }

  isNegative(): boolean {
    return djs.isNegative(this._raw);
  }

  isZero(): boolean {
    return djs.isZero(this._raw);
  }

  /**
   * Adds b to the current object and returns a new one with the result
   *
   * @param b - Money to be added
   * @returns - Money with result
   */
  add(b: Money): Money {
    const result = djs.toSnapshot(djs.add(b.raw, this.raw));
    return new Money(result.amount, this.currency, result.scale);
  }

  /**
   * Subtracts b to the current object and returns a new one with the result
   *
   * @param b - Money to be subtracted
   * @returns - Money with result
   */
  subtract(b: Money): Money {
    const result = djs.toSnapshot(djs.subtract(this.raw, b.raw));
    return new Money(result.amount, this.currency, result.scale);
  }

  /**
   * Multiplies b to the current object and returns a new one with the result
   * Initially, we try to multiply without transforming b but sometimes dinerojs
   * returns that it is an invalid amount (I have no idea why). If it fails, then
   * we transform b to `amountWithScale` and multiply with that.
   *
   * @param b - amount to be multiplied to in scale 0
   * @returns - Money with result
   */
  multiply(b: number): Money {
    let result;

    try {
      result = djs.multiply(this.raw, b);
    } catch {
      result = djs.multiply(this.raw, toAmountWithScale(b));
    }

    const snapshot = djs.toSnapshot(result);
    return new Money(snapshot.amount, this.currency, snapshot.scale);
  }

  /**
   * Checks for equality with b by comparing amount and currency
   *
   * @returns - boolean
   */
  equals(b: Money): boolean {
    return djs.equal(this._raw, b._raw);
  }

  /**
   * Converts a Money object to the currency specified
   *
   * @param to - Money object containing the rate to
   */
  convert(
    to: string,
    rate: number,
  ): Money {
    const rates = {
      [to]: {
        ...toAmountWithScale(rate),
      },
    };
    const currency = toCurrency(to);

    try {
      // @ts-ignore
      const result = djs.toSnapshot(djs.convert(this.raw, currency, rates));
      return new Money(result.amount, currency.code, result.scale);
    } catch (e) {
      throw new Error(`Error converting ${this.format()} to ${to} with rates ${JSON.stringify(rates)}: ${e}`);
    }
  }
}

function toCurrency(currency: string): DineroCurrency<number> {
  // @ts-ignore
  let dineroCurrency = currencies[currency];
  if (!dineroCurrency) {
    dineroCurrency = {
      code: currency,
      base: 10,
      exponent: 2,
    };
  }
  return dineroCurrency;
}
