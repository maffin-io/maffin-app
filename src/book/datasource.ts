import initSqlJs from 'sql.js';
import { DataSource } from 'typeorm';
import { useSWRConfig } from 'swr';

import {
  Account,
  Book,
  Commodity,
  Price,
  Split,
  Transaction,
} from './entities';

export async function initDB(database: Uint8Array): Promise<DataSource> {
  await initSqlJs({
    // Required to load the wasm binary asynchronously. Of course, you can host it wherever you want
    locateFile: file => `https://sql.js.org/dist/${file}`,
  });

  let datasource: DataSource;

  if (database.length !== 0) {
    datasource = new DataSource({
      type: 'sqljs',
      synchronize: true,
      database,
      logging: false,
      entities: [Account, Book, Commodity, Price, Split, Transaction],
    });
    await datasource.initialize();
  } else {
    datasource = new DataSource({
      type: 'sqljs',
      synchronize: true,
      logging: false,
      entities: [Account, Book, Commodity, Price, Split, Transaction],
    });
    await datasource.initialize();
    await createEmptyBook();
  }

  return datasource;
}

// TODO: Need to mutate the accounts SWR key so Root is shown!
async function createEmptyBook() {
  const rootAccount = Account.create({
    guid: 'rootAccount',
    name: 'Root',
    type: 'ROOT',
  });
  await Account.upsert(rootAccount, ['guid']);

  await Book.upsert(
    {
      guid: 'maffinBook',
      fk_root: rootAccount.guid,
    },
    ['guid'],
  );
}
