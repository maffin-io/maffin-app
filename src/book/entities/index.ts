export { default as Account } from './Account';
export { default as BankConfig } from './BankConfig';
export { default as Commodity } from './Commodity';
export { default as Price } from './Price';
export { default as Split } from './Split';
export { default as Transaction } from './Transaction';
export { default as Book } from './Book';
export { default as BaseEntity } from './BaseEntity';
