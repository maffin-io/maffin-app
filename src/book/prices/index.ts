export { default as PriceDBMap } from './PriceDBMap';
