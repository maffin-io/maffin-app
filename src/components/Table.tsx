import React from 'react';
import {
  ColumnDef,
  TableOptions,
  useReactTable,
  flexRender,
  getCoreRowModel,
  getSortedRowModel,
  getExpandedRowModel,
  ColumnSort,
  ExpandedState,
  getPaginationRowModel,
} from '@tanstack/react-table';
import classNames from 'classnames';
import { FaSortDown, FaSortUp } from 'react-icons/fa';

import Pagination from '@/components/Pagination';

export type TableProps<T extends object> = {
  id: string,
  columns: ColumnDef<T>[],
  data: T[],
  initialSort?: ColumnSort,
  showHeader?: boolean,
  showPagination?: boolean,
  tdClassName?: string,
  isExpanded?: boolean,
} & Partial<TableOptions<T>>;

export default function Table<T extends object = {}>(
  {
    id,
    initialSort,
    showHeader = true,
    showPagination = false,
    tdClassName = 'px-6 py-4',
    isExpanded = false,
    ...props
  }: TableProps<T>,
): JSX.Element {
  const tableConfig: TableOptions<T> = {
    getCoreRowModel: getCoreRowModel(),
    getSortedRowModel: getSortedRowModel(),
    getExpandedRowModel: getExpandedRowModel(),
    enableExpanding: true,
    initialState: {
      ...props.initialState,
      sorting: (initialSort && [initialSort]) || undefined,
      expanded: isExpanded as ExpandedState,
    },
    ...props,
  };

  if (showPagination && !props.manualPagination) {
    tableConfig.getPaginationRowModel = getPaginationRowModel();
  }

  const table = useReactTable(tableConfig);

  return (
    <>
      <div className="relative overflow-hidden rounded-md">
        <table id={id} className="w-full text-sm text-left">
          {
            showHeader
            && (
              <thead className="bg-white dark:bg-dark-700">
                {table.getHeaderGroups().map((headerGroup) => (
                  <tr key={headerGroup.id}>
                    {headerGroup.headers.map(header => (
                      <th
                        key={header.id}
                        scope="col"
                        className="px-6 py-3"
                        onClick={header.column.getToggleSortingHandler()}
                      >
                        <div className="flex">
                          {flexRender(header.column.columnDef.header, header.getContext())}
                          {(
                            header.column.columnDef.header !== ''
                              && header.column.getCanSort()
                              && (
                                <div className="flex items-center cursor-pointer">
                                  <FaSortUp
                                    className={
                                      classNames(
                                        {
                                          'opacity-100': header.column.getIsSorted() === 'asc',
                                          'opacity-50': header.column.getIsSorted() !== 'asc',
                                        },
                                        'absolute',
                                      )
                                    }
                                  />
                                  <FaSortDown
                                    className={
                                      classNames(
                                        {
                                          'opacity-100': header.column.getIsSorted() === 'desc',
                                          'opacity-50': header.column.getIsSorted() !== 'desc',
                                        },
                                        'absolute',
                                      )
                                    }
                                  />
                                </div>
                              )
                          )}
                        </div>
                      </th>
                    ))}
                  </tr>
                ))}
              </thead>
            )
          }

          <tbody>
            {table.getRowModel().rows.map(row => (
              <tr key={row.id} className="border-b border-white/70 dark:border-dark-700">
                {row.getVisibleCells().map(cell => (
                  <td key={cell.id} className={tdClassName}>
                    {flexRender(cell.column.columnDef.cell, cell.getContext())}
                  </td>
                ))}
              </tr>
            ))}
          </tbody>
        </table>
      </div>

      {
        showPagination
        && <Pagination<T> table={table} />
      }
    </>
  );
}
