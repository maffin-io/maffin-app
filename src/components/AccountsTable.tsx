import React from 'react';
import { ColumnDef } from '@tanstack/react-table';
import Link from 'next/link';
import { BiCircle, BiSolidRightArrow, BiSolidDownArrow } from 'react-icons/bi';
import classNames from 'classnames';

import Table from '@/components/Table';
import type { AccountsTree } from '@/types/accounts';

export type AccountsTableProps = {
  tree: AccountsTree,
};

export default function AccountsTable(
  {
    tree,
  }: AccountsTableProps,
): JSX.Element {
  return (
    <Table<AccountsTree>
      columns={columns}
      data={tree.children}
      initialSort={{ id: 'total', desc: true }}
      showHeader={false}
      showPagination={false}
      tdClassName="p-2"
      getSubRows={row => row.children}
    />
  );
}

const columns: ColumnDef<AccountsTree>[] = [
  {
    accessorKey: 'name',
    id: 'name',
    enableSorting: false,
    header: '',
    cell: ({ row }) => (
      <div
        className={`flex pl-${row.depth * 4}`}
      >
        {row.getCanExpand() ? (
          <button
            type="button"
            onClick={() => row.toggleExpanded()}
          >
            {(
              row.getIsExpanded()
                ? <BiSolidDownArrow className="mr-1 text-xs opacity-20" />
                : <BiSolidRightArrow className="mr-1 text-xs opacity-20" />
            )}
          </button>
        ) : (
          <button
            type="button"
            className="cursor-default"
          >
            <BiCircle className="mr-1 text-xs opacity-20" />
          </button>
        )}
        <Link
          className={classNames('badge hover:text-slate-300', {
            'bg-green-500/20 text-green-300': row.original.account.type === 'INCOME',
            'bg-red-500/20 text-red-300': row.original.account.type === 'EXPENSE',
            'bg-cyan-500/20 text-cyan-300': ['ASSET', 'BANK'].includes(row.original.account.type),
            'bg-orange-500/20 text-orange-300': row.original.account.type === 'LIABILITY',
            'bg-violet-500/20 text-violet-300': ['STOCK', 'MUTUAL'].includes(row.original.account.type),
          })}
          href={`/dashboard/accounts/${row.original.account.guid}`}
        >
          {row.original.account.name}
        </Link>
      </div>
    ),
  },
  {
    accessorFn: (row: AccountsTree) => row.total.toNumber(),
    id: 'total',
    header: '',
    cell: ({ row }) => (
      <span>
        {row.original.total.format()}
      </span>
    ),
  },
];
