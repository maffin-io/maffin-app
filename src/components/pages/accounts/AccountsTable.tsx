import React from 'react';
import { DateTime } from 'luxon';
import { ColumnDef } from '@tanstack/react-table';
import Link from 'next/link';
import { BiCircle, BiSolidRightArrow, BiSolidDownArrow } from 'react-icons/bi';
import classNames from 'classnames';
import { Tooltip } from 'react-tooltip';

import Money from '@/book/Money';
import Table from '@/components/Table';
import type { AccountsMap } from '@/types/book';
import { MonthlyTotals } from '@/lib/queries';
import { Account } from '@/book/entities';
import {
  isInvestment,
  isAsset,
  isLiability,
} from '@/book/helpers/accountType';
import { useAccountsTotals, useAccounts } from '@/hooks/api';
import mapAccounts from '@/helpers/mapAccounts';

export type AccountsTableProps = {
  selectedDate?: DateTime,
  isExpanded?: boolean,
};

type AccountsTableRow = {
  account: Account,
  total: Money,
  leaves: AccountsTableRow[],
};

export default function AccountsTable(
  {
    selectedDate = DateTime.now(),
    isExpanded = false,
  }: AccountsTableProps,
): JSX.Element {
  const { data } = useAccounts();
  const { data: monthlyTotals } = useAccountsTotals();

  const accounts = mapAccounts(data);
  const tree = getTreeTotals(accounts.type_root, accounts, monthlyTotals || {}, selectedDate);

  return (
    <Table<AccountsTableRow>
      id="accounts-table"
      columns={columns}
      data={tree.leaves}
      initialState={{
        sorting: [{ id: 'total', desc: true }],
      }}
      showHeader={false}
      tdClassName="p-2"
      getSubRows={row => row.leaves}
      isExpanded={isExpanded}
    />
  );
}

function getTreeTotals(
  current: Account,
  accounts: AccountsMap,
  monthlyTotals: MonthlyTotals,
  selectedDate: DateTime,
): AccountsTableRow {
  const leaves: AccountsTableRow[] = [];
  current.childrenIds.forEach(childId => {
    const childAccount = accounts[childId];
    if (!childAccount.hidden && childAccount.parentId === current.guid) {
      leaves.push(getTreeTotals(childAccount, accounts, monthlyTotals, selectedDate));
    }
  });

  let accountTotal = (monthlyTotals[current.guid] || {})[selectedDate.toFormat('MM/yyyy')]
    || new Money(0, current.commodity?.mnemonic || '');
  if (!isAsset(current) && !isLiability(current)) {
    accountTotal = Object.entries(monthlyTotals[current.guid] || {}).reduce(
      (total, [monthYear, amount]) => {
        if (DateTime.fromFormat(monthYear, 'MM/yyyy') <= selectedDate) {
          return total.add(amount);
        }
        return total;
      },
      new Money(0, current.commodity?.mnemonic || ''),
    );
  }

  return {
    account: current,
    total: current.type === 'INCOME' || isLiability(current) ? accountTotal.multiply(-1) : accountTotal,
    leaves,
  };
}

const columns: ColumnDef<AccountsTableRow>[] = [
  {
    accessorKey: 'name',
    id: 'name',
    enableSorting: false,
    header: '',
    cell: ({ row }) => (
      <div
        className={`flex pl-${row.depth * 4}`}
      >
        {row.getCanExpand() ? (
          <button
            type="button"
            onClick={() => row.toggleExpanded()}
          >
            {(
              row.getIsExpanded()
                ? <BiSolidDownArrow className="mr-1 text-xs opacity-50" />
                : <BiSolidRightArrow className="mr-1 text-xs opacity-50" />
            )}
          </button>
        ) : (
          <button
            type="button"
            className="cursor-default"
          >
            <BiCircle className="mr-1 text-xs opacity-50" />
          </button>
        )}
        {
          (
            row.original.account.placeholder
            && (
              <span
                data-tooltip-id={row.original.account.guid}
                className={classNames('badge cursor-default', {
                  success: row.original.account.type === 'INCOME',
                  danger: row.original.account.type === 'EXPENSE',
                  info: isAsset(row.original.account),
                  warning: isLiability(row.original.account),
                  misc: isInvestment(row.original.account),
                })}
              >
                {row.original.account.name}
              </span>
            )
          ) || (
            <Link
              data-tooltip-id={row.original.account.guid}
              className={classNames('badge hover:text-slate-300', {
                success: row.original.account.type === 'INCOME',
                danger: row.original.account.type === 'EXPENSE',
                info: isAsset(row.original.account),
                warning: isLiability(row.original.account),
                misc: isInvestment(row.original.account),
              })}
              href={`/dashboard/accounts/${row.original.account.guid}`}
            >
              {row.original.account.name}
            </Link>
          )
        }

        {
          row.original.account.description
          && (
            <Tooltip
              id={row.original.account.guid}
              className="tooltip"
              disableStyleInjection
            >
              {row.original.account.description}
            </Tooltip>
          )
        }
      </div>
    ),
  },
  {
    accessorFn: (row: AccountsTableRow) => row.total.toNumber(),
    id: 'total',
    header: '',
    cell: ({ row }) => (
      <span>
        {row.original.total.format()}
      </span>
    ),
  },
];
