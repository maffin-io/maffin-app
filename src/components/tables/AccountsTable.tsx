import React from 'react';
import { ColumnDef } from '@tanstack/react-table';
import Link from 'next/link';
import { BiCircle, BiSolidRightArrow, BiSolidDownArrow } from 'react-icons/bi';

import Money from '@/book/Money';
import { Tooltip } from '@/components/tooltips';
import Table from '@/components/tables/Table';
import type { AccountsMap } from '@/types/book';
import { Account } from '@/book/entities';
import { useAccounts, useAccountsTotals } from '@/hooks/api';
import mapAccounts from '@/helpers/mapAccounts';
import { accountColorCode } from '@/helpers/classNames';

export type AccountsTableProps = {
  guids: string[],
  isExpanded?: boolean,
};

export default function AccountsTable(
  {
    guids,
    isExpanded = false,
  }: AccountsTableProps,
): JSX.Element {
  const { data } = useAccounts();
  const { data: accountsTotal } = useAccountsTotals();

  const accounts = mapAccounts(data);
  const trees: AccountsTableRow[] = [];
  guids.forEach(guid => {
    if (accounts[guid] && !accounts[guid].hidden) {
      trees.push(getTree(
        accounts[guid],
        accounts,
        accountsTotal || {},
      ));
    }
  });

  return (
    <Table<AccountsTableRow>
      id={`${guids.join('-')}-table`}
      columns={columns}
      data={trees}
      initialState={{
        sorting: [{ id: 'total', desc: true }],
      }}
      showHeader={false}
      tdClassName="p-2"
      getSubRows={row => row.leaves}
      isExpanded={isExpanded}
    />
  );
}

function getTree(
  current: Account,
  accounts: AccountsMap,
  accountsTotal: { [guid: string]: Money },
): AccountsTableRow {
  const leaves: AccountsTableRow[] = [];
  current.childrenIds.forEach(childId => {
    const childAccount = accounts[childId];
    if (!childAccount.hidden && childAccount.parentId === current.guid) {
      leaves.push(getTree(childAccount, accounts, accountsTotal));
    }
  });

  const accountTotal = accountsTotal[current.guid] || new Money(0, current.commodity?.mnemonic || '');

  return {
    account: current,
    total: accountTotal,
    leaves,
  };
}

const columns: ColumnDef<AccountsTableRow>[] = [
  {
    accessorKey: 'name',
    id: 'name',
    enableSorting: false,
    header: '',
    cell: ({ row }) => (
      <div
        className={`flex pl-${row.depth * 4}`}
      >
        {row.getCanExpand() ? (
          <button
            type="button"
            onClick={() => row.toggleExpanded()}
          >
            {(
              row.getIsExpanded()
                ? <BiSolidDownArrow className="mr-1 text-xs opacity-50" />
                : <BiSolidRightArrow className="mr-1 text-xs opacity-50" />
            )}
          </button>
        ) : (
          <button
            type="button"
            className="cursor-default"
          >
            <BiCircle className="mr-1 text-xs opacity-50" />
          </button>
        )}
        <Link
          data-tooltip-id={row.original.account.guid}
          className={accountColorCode(row.original.account, 'badge hover:text-white')}
          href={`/dashboard/accounts/${row.original.account.guid}`}
        >
          {row.original.account.name}
        </Link>

        {
          row.original.account.description
          && (
            <Tooltip
              place="bottom"
              className="max-w-[80%]"
              id={row.original.account.guid}
            >
              {row.original.account.description}
            </Tooltip>
          )
        }
      </div>
    ),
  },
  {
    accessorFn: (row: AccountsTableRow) => row.total.toNumber(),
    id: 'total',
    header: '',
    cell: ({ row }) => (
      <span>
        {row.original.total.format()}
      </span>
    ),
  },
];

type AccountsTableRow = {
  account: Account,
  total: Money,
  leaves: AccountsTableRow[],
};
