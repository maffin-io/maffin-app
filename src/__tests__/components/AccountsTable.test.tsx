import React from 'react';
import { render, screen } from '@testing-library/react';

import AccountsTable, { Record } from '@/components/AccountsTable';
import Table from '@/components/Table';
import { PriceDBMap } from '@/book/prices';
import { Account } from '@/book/entities';
import Money from '@/book/Money';

jest.mock('@/components/Table', () => jest.fn(
  () => <div data-testid="Table" />,
));
const TableMock = Table as jest.MockedFunction<typeof Table>;

describe('AccountsTable', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('creates empty Table with expected params', async () => {
    const { container } = render(
      <AccountsTable
        accounts={[{} as Account]}
        todayPrices={new PriceDBMap()}
      />,
    );

    await screen.findByTestId('Table');
    expect(Table).toHaveBeenLastCalledWith(
      {
        columns: [
          {
            header: '',
            id: 'name',
            enableSorting: false,
            accessorKey: 'name',
            cell: expect.any(Function),
          },
          {
            header: '',
            id: 'total',
            accessorFn: expect.any(Function),
            cell: expect.any(Function),
          },
        ],
        data: [],
        initialSort: {
          desc: true,
          id: 'total',
        },
        showHeader: false,
        showPagination: false,
      },
      {},
    );
    expect(container).toMatchSnapshot();
  });

  it('builds subRows as expected with non investment accounts', async () => {
    const accounts = [
      {
        guid: 'root',
        name: 'Root',
        type: 'ROOT',
        childrenIds: ['a1'],
      } as Account,
      {
        guid: 'a1',
        name: 'Assets',
        total: new Money(10, 'EUR'),
        commodity: {
          mnemonic: 'EUR',
        },
        type: 'ASSET',
        childrenIds: ['a2'],
      } as Account,
      {
        guid: 'a2',
        name: 'Bank',
        total: new Money(1000, 'USD'),
        commodity: {
          mnemonic: 'USD',
        },
        type: 'BANK',
        childrenIds: [] as string[],
      } as Account,
    ];

    render(
      <AccountsTable
        accounts={accounts}
        todayPrices={{
          getPrice: (from, to, date) => ({
            guid: `${from}.${to}.${date}`,
            value: 0.98,
          }),
        } as PriceDBMap}
      />,
    );

    await screen.findByTestId('Table');
    expect(Table).toBeCalledTimes(1);
    expect(Table).toHaveBeenLastCalledWith(
      {
        columns: [
          {
            header: '',
            id: 'name',
            enableSorting: false,
            accessorKey: 'name',
            cell: expect.any(Function),
          },
          {
            header: '',
            id: 'total',
            accessorFn: expect.any(Function),
            cell: expect.any(Function),
          },
        ],
        data: [
          {
            guid: 'a1',
            name: 'Assets',
            type: 'ASSET',
            total: expect.any(Money),
            subRows: [
              {
                guid: 'a2',
                name: 'Bank',
                type: 'BANK',
                total: expect.any(Money),
                subRows: [],
              },
            ],
          },
        ],
        initialSort: {
          desc: true,
          id: 'total',
        },
        showHeader: false,
        showPagination: false,
      },
      {},
    );

    const data = TableMock.mock.calls[0][0].data as Record[];
    expect(data[0].subRows[0].total.toString()).toEqual('1000.00 USD');
    expect(data[0].total.toString()).toEqual('990.00 EUR'); // 1000 USD * 0.98 + 10 EUR
  });

  it('builds subRows as expected with investment accounts', async () => {
    const accounts = [
      {
        guid: 'root',
        name: 'Root',
        type: 'ROOT',
        childrenIds: ['a1'],
      } as Account,
      {
        guid: 'a1',
        name: 'Stocks',
        total: new Money(0, 'EUR'),
        commodity: {
          mnemonic: 'EUR',
        },
        type: 'ASSET',
        childrenIds: ['a2'],
      } as Account,
      {
        guid: 'a2',
        name: 'GOOGL',
        total: new Money(2, 'GOOGL'),
        commodity: {
          mnemonic: 'GOOGL',
        },
        type: 'STOCK',
        childrenIds: [] as string[],
      } as Account,
    ];

    render(
      <AccountsTable
        accounts={accounts}
        todayPrices={{
          getPrice: (from, to, date) => ({
            guid: `${from}.${to}.${date}`,
            value: 0.98,
          }),
          getStockPrice: (from, date) => ({
            guid: `${from}.${date}`,
            value: 100,
            currency: {
              mnemonic: 'USD',
            },
          }),
        } as PriceDBMap}
      />,
    );

    await screen.findByTestId('Table');
    expect(Table).toBeCalledTimes(1);
    expect(Table).toHaveBeenLastCalledWith(
      {
        columns: [
          {
            header: '',
            id: 'name',
            enableSorting: false,
            accessorKey: 'name',
            cell: expect.any(Function),
          },
          {
            header: '',
            id: 'total',
            accessorFn: expect.any(Function),
            cell: expect.any(Function),
          },
        ],
        data: [
          {
            guid: 'a1',
            name: 'Stocks',
            type: 'ASSET',
            total: expect.any(Money),
            subRows: [
              {
                guid: 'a2',
                name: 'GOOGL',
                type: 'STOCK',
                total: expect.any(Money),
                subRows: [],
              },
            ],
          },
        ],
        initialSort: {
          desc: true,
          id: 'total',
        },
        showHeader: false,
        showPagination: false,
      },
      {},
    );

    const data = TableMock.mock.calls[0][0].data as Record[];
    expect(data[0].subRows[0].total.toString()).toEqual('200.00 USD'); // 2 GOOGL * 100 USD
    expect(data[0].total.toString()).toEqual('196.00 EUR'); // 200 USD * 0.98
  });

  it('renders Name column as expected', async () => {
    const account = {
      guid: 'a1',
      name: 'Assets',
      total: new Money(10, 'EUR'),
      commodity: {
        mnemonic: 'EUR',
      },
      type: 'ASSET',
      childrenIds: [] as string[],
    } as Account;

    render(
      <AccountsTable
        accounts={[account]}
        todayPrices={new PriceDBMap()}
      />,
    );

    await screen.findByTestId('Table');
    expect(Table).toBeCalledTimes(1);
    const nameCol = TableMock.mock.calls[0][0].columns[0];

    expect(nameCol.cell).not.toBeUndefined();
    const { container } = render(
      // @ts-ignore
      nameCol.cell({
        row: {
          original: account,
        },
      }),
    );

    await screen.findByText('Assets');
    expect(container).toMatchSnapshot();
  });

  it('renders Total column as expected', async () => {
    const account = {
      guid: 'a1',
      name: 'Assets',
      total: new Money(10, 'EUR'),
      commodity: {
        mnemonic: 'EUR',
      },
      type: 'ASSET',
      childrenIds: [] as string[],
    } as Account;

    render(
      <AccountsTable
        accounts={[account]}
        todayPrices={new PriceDBMap()}
      />,
    );

    await screen.findByTestId('Table');
    expect(Table).toBeCalledTimes(1);
    const totalCol = TableMock.mock.calls[0][0].columns[1];

    expect(
      // @ts-ignore
      totalCol.accessorFn({ total: new Money(10, 'EUR') }),
    ).toEqual(10);

    expect(totalCol.cell).not.toBeUndefined();
    const { container } = render(
      // @ts-ignore
      totalCol.cell({
        row: {
          original: account,
        },
      }),
    );

    await screen.findByText('10.00 â‚¬');
    expect(container).toMatchSnapshot();
  });
});
