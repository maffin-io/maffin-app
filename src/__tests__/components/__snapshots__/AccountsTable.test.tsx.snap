// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountsTable creates empty Table with expected params 1`] = `
<div>
  <div
    data-testid="Table"
  />
</div>
`;

exports[`AccountsTable renders Name column as expected 1`] = `
<div>
  <a
    href="/dashboard/accounts/a1"
  >
    Assets
  </a>
</div>
`;

exports[`AccountsTable renders Total column as expected 1`] = `
<div>
  <span>
    10.00 â‚¬
  </span>
</div>
`;
